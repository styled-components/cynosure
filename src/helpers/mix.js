// @flow

import endsWith from '../internalHelpers/_endsWith'
import checkFix from '../internalHelpers/_checkFix'
import Color from '../internalHelpers/_Color'
import stripUnit from '../helpers/stripUnit'

type mixColor = {
  color1: string,
  color2: string,
  weight: number,
}

/**
 * Mixes two colors together. Specifically, takes the average of each of the
 * RGB components, optionally weighted by the given percentage. The opacity
 * of the colors is also considered when weighting the components.
 *
 * The weight specifies the amount of the first color that should be included
 * in the returned color. The default, `50%`, means that half the first color
 * and half the second color should be used. `25%` means that a quarter of
 * the first color and three quarters of the second color should be used.
 *
 * @example
 * // Styles as object usage
 *
 * const styles = {
 *   ...color: mix('#f00', '#00f')
 *   ...color: mix('#f00', '#00f', 25)
 *   ...color: mix('#f00', '#00f', '25%')
 * }
 *
 *
 * // styled-components usage
 * const div = styled.div`
 *   color: ${mix('#f00', '#00f')};
 *   color: ${mix('#f00', '#00f', 25)};
 *   color: ${mix('#f00', '#00f', '25%')};
 *`
 *
 * // CSS in JS Output
 *
 * div {
 *   color: "#7f007f";
 *   color: "#3f00bf";
 *   color: "#3f00bf";
 * }
 */
function mix(color1 : mixColor | string, color2: string, weight: string|number = 50):string {
  const newColor1 = checkFix(color1)
  const newColor2 = checkFix(color2)
  let newWeight = weight
  if (typeof weight === 'string') {
    if (endsWith(weight, '%')) {
      newWeight = stripUnit(weight)// newWeight.substr(-1)
    }
  }
  if (newWeight > 100) newWeight = 100

  const c1 = new Color(newColor1)
  const c2 = new Color(newColor2)
  const newColor = new Color('mix', [c1, c2, newWeight])
  return newColor.hex()
}

export default mix
