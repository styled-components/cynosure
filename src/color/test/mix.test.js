// @flow
import mix from '../mix'

describe('mix', () => {
  it('should mix two colors with by a weight of 25%', () => {
    expect(mix(0.25, '#f00', '#00f')).toMatchSnapshot()
  })

  it('should mix a color with an 8-digit hex color', () => {
    expect(mix(0.5, '#FF00007F', '#00f')).toMatchSnapshot()
  })

  it('should mix a 8-digit hex color with a 4-digit hex color', () => {
    expect(mix(0.5, '#FF00007F', '#0f08')).toMatchSnapshot()
  })

  it('should mix a color with a color with an opacity lower than 1', () => {
    expect(mix(0.51, 'rgba(242, 236, 228, 0.5)', '#6b717f')).toMatchSnapshot()
  })

  it('should mix two rgba colors', () => {
    expect(
      mix(0.7, 'rgba(0, 0, 0, 1)', 'rgba(255, 255, 255, 0)'),
    ).toMatchSnapshot()
  })

  it('should mix two colors when weight is a string', () => {
    expect(mix('0.5', 'rgba(255, 0, 0, 0.5)', '#00f')).toMatchSnapshot()
  })

  it('should return otherColor when passed transparent color', () => {
    expect(mix('0.5', 'transparent', '#00f')).toMatchSnapshot()
  })

  it('should return color when passed transparent otherColor', () => {
    expect(mix('0.5', 'rgba(255, 0, 0, 0.5)', 'transparent')).toMatchSnapshot()
  })

  it('should return transparent when passed transparent for both colors', () => {
    expect(mix('0.5', 'transparent', 'transparent')).toMatchSnapshot()
  })

  it('should return the second color when weight is 0', () => {
    expect(
      mix(0, 'rgba(0, 0, 0, 1)', 'rgba(255, 255, 255, 0)'),
    ).toMatchSnapshot()
  })
})
