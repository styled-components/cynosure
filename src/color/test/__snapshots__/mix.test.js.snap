// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mix should mix a 8-digit hex color with a 4-digit hex color 1`] = `"rgba(123,131,0,0.515)"`;

exports[`mix should mix a color with a color with an opacity lower than 1 1`] = `"rgba(141,144,153,0.745)"`;

exports[`mix should mix a color with an 8-digit hex color 1`] = `"rgba(63,0,191,0.75)"`;

exports[`mix should mix two colors when weight is a string 1`] = `"rgba(63,0,191,0.75)"`;

exports[`mix should mix two colors with by a weight of 25% 1`] = `"#3f00bf"`;

exports[`mix should mix two rgba colors 1`] = `"rgba(0,0,0,0.7)"`;

exports[`mix should return color when passed transparent otherColor 1`] = `"rgba(255, 0, 0, 0.5)"`;

exports[`mix should return otherColor when passed transparent color 1`] = `"#00f"`;

exports[`mix should return the second color when weight is 0 1`] = `"rgba(255, 255, 255, 0)"`;

exports[`mix should return transparent when passed transparent for both colors 1`] = `"transparent"`;
